game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java

```diff
@@ -0,0 +1,143 @@
+   * @param targetUnits Collection of units that are being hit
+   * @return List of FiringGroup
+   */
+  public static List<FiringGroup> groupBySuicideOnHit(
+      final String name, final Collection<Unit> firingUnits, final Collection<Unit> targetUnits) {
+    final List<FiringGroup> groups = new ArrayList<>();
+    final List<Collection<Unit>> separatedBySuicide = separateSuicideOnHit(firingUnits);
+
+    // ensure each firing group has a unique name by adding prefixes
+    // if the firing groups have different types of suicide units
+    if (separatedBySuicide.size() == 1) {
+      groups.add(
+          new FiringGroup(
+              name,
+              name,
+              separatedBySuicide.get(0),
+              targetUnits,
+              Matches.unitIsSuicideOnHit().test(separatedBySuicide.get(0).iterator().next())));
+
+    } else if (separatedBySuicide.size() == 2
+        && Matches.unitIsSuicideOnHit().test(separatedBySuicide.get(0).iterator().next())
+            != Matches.unitIsSuicideOnHit().test(separatedBySuicide.get(1).iterator().next())) {
+      groups.addAll(
+          generateFiringGroupsWithOneSuicideAndOneNonSuicide(
+              name, targetUnits, separatedBySuicide));
+
+    } else {
+      groups.addAll(generateFiringGroups(name, targetUnits, separatedBySuicide));
+    }
+    return groups;
+  }
+
+  /**
+   * Separate the suicide on hit units from the others and group them by their type. The suicide on
+   * hit units need to fire separately so that they can be removed if they hit.
+   */
+  private static List<Collection<Unit>> separateSuicideOnHit(final Collection<Unit> units) {
+
+    final Map<UnitType, Collection<Unit>> map = new HashMap<>();
+    for (final Unit unit : CollectionUtils.getMatches(units, Matches.unitIsSuicideOnHit())) {
+      final UnitType type = unit.getType();
+      if (map.containsKey(type)) {
+        map.get(type).add(unit);
+      } else {
+        final Collection<Unit> unitList = new ArrayList<>();
+        unitList.add(unit);
+        map.put(type, unitList);
+      }
+    }
+
+    final List<Collection<Unit>> result = new ArrayList<>(map.values());
+    final Collection<Unit> remainingUnits =
+        CollectionUtils.getMatches(units, Matches.unitIsSuicideOnHit().negate());
+    if (!remainingUnits.isEmpty()) {
+      result.add(remainingUnits);
+    }
+    return result;
+  }
+
+  /** Handle the case where there are only two groups and one is suicide and the other is not */
+  private static List<FiringGroup> generateFiringGroupsWithOneSuicideAndOneNonSuicide(
+      final String name,
+      final Collection<Unit> targetUnits,
+      final List<Collection<Unit>> separatedBySuicide) {
+    final List<FiringGroup> groups = new ArrayList<>();
+    for (final Collection<Unit> newFiringUnits : separatedBySuicide) {
+      final boolean isSuicideOnHit =
+          Matches.unitIsSuicideOnHit().test(newFiringUnits.iterator().next());
```

Comment:
Will we always be guaranteed to have a first unit in: `newFiringUnits.iterator().next()`?
I do not know, maybe there is some code earlier that guarantees this, however it seems like a pretty strong precondition that is not necessarily documented (hopefully enforced). If we convert the incoming list to a map, that would resolve this problem as you would not need to check any of the unit elements to determine if they are 'suicide on hit', you could retrieve them directly from the input map.